docs/manpages/vfs_zfsacl.8
'\" t
.\"     Title: vfs_zfsacl
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 07/02/2020
.\"    Manual: System Administration tools
.\"    Source: Samba 4.12.5
.\"  Language: English
.\"
.TH "VFS_ZFSACL" "8" "07/02/2020" "Samba 4\&.12\&.5" "System Administration tools"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
vfs_zfsacl \- ZFS ACL samba module
.SH "SYNOPSIS"
.HP \w'\ 'u
vfs objects = zfsacl
.SH "DESCRIPTION"
.PP
This VFS module is part of the
\fBsamba\fR(7)
suite\&.
.PP
The
zfsacl
VFS module is the home for all ACL extensions that Samba requires for proper integration with ZFS\&.
.PP
Currently the zfsacl vfs module provides extensions in following areas :
.RS
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
NFSv4 ACL Interfaces with configurable options for ZFS
.RE
.sp
.RE
.PP
NOTE:This module follows the posix\-acl behaviour and hence allows permission stealing via chown\&. Samba might allow at a later point in time, to restrict the chown via this module as such restrictions are the responsibility of the underlying filesystem than of Samba\&.
.PP
This module makes use of the smb\&.conf parameter
\m[blue]\fBacl map full control = acl map full control\fR\m[]
When set to yes (the default), this parameter will add in the FILE_DELETE_CHILD bit on a returned ACE entry for a file (not a directory) that already contains all file permissions except for FILE_DELETE and FILE_DELETE_CHILD\&. This can prevent Windows applications that request GENERIC_ALL access from getting ACCESS_DENIED errors when running against a filesystem with NFSv4 compatible ACLs\&.
.PP
This module is stackable\&.
.PP
Since Samba 4\&.0 all options are per share options\&.
.SH "OPTIONS"
.PP
nfs4:mode = [ simple | special ]
.RS 4
Controls substitution of special IDs (OWNER@ and GROUP@) on ZFS\&. The use of mode simple is recommended\&. In this mode only non inheriting ACL entries for the file owner and group are mapped to special IDs\&.
.sp
The following MODEs are understood by the module:
.RS
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
simple(default)
\- use OWNER@ and GROUP@ special IDs for non inheriting ACEs only\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
special(deprecated)
\- use OWNER@ and GROUP@ special IDs in ACEs for all file owner and group ACEs\&.
.RE
.sp
.RE
.RE
.PP
nfs4:acedup = [dontcare|reject|ignore|merge]
.RS 4
This parameter configures how Samba handles duplicate ACEs encountered in ZFS ACLs\&. ZFS allows/creates duplicate ACE for different bits for same ID\&.
.sp
Following is the behaviour of Samba for different values :
.RS
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
dontcare (default)
\- copy the ACEs as they come
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
reject (deprecated)
\- stop operation and exit with error on ACL set op
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
ignore (deprecated)
\- don\*(Aqt include the second matching ACE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
merge
\- bitwise OR the 2 ace\&.flag fields and 2 ace\&.mask fields of the 2 duplicate ACEs into 1 ACE
.RE
.sp
.RE
.RE
.PP
nfs4:chown = [yes|no]
.RS 4
This parameter allows enabling or disabling the chown supported by the underlying filesystem\&. This parameter should be enabled with care as it might leave your system insecure\&.
.sp
Some filesystems allow chown as a) giving b) stealing\&. It is the latter that is considered a risk\&.
.sp
Following is the behaviour of Samba for different values :
.RS
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
yes
\- Enable chown if as supported by the under filesystem
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
no (default)
\- Disable chown
.RE
.sp
.RE
.RE
.PP
zfsacl:denymissingspecial = [yes|no]
.RS 4
Prevent users from setting an ACL that lacks NFSv4 special entries (owner@, group@, everyone@)\&. ZFS will automatically generate these these entries when calculating the inherited ACL of new files if the ACL of the parent directory lacks an inheriting special entry\&. This may result in user confusion and unexpected change in permissions of files and directories as the inherited ACL is generated\&.
.RS
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
yes
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
no (default)
.RE
.sp
.RE
.RE
.PP
zfsacl:map_dacl_protected = [yes|no]
.RS 4
If enabled and the ZFS ACL on the underlying filesystem does not contain any inherited access control entires, then set the SEC_DESC_DACL_PROTECTED flag on the Security Descriptor returned to SMB clients\&. This ensures correct Windows client behavior when disabling inheritance on directories\&.
.sp
Following is the behaviour of Samba for different values :
.RS
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
yes
\- Enable mapping to SEC_DESC_DACL_PROTECTED
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
no (default)
.RE
.sp
.RE
.RE
.SH "EXAMPLES"
.PP
A ZFS mount can be exported via Samba as follows :
.sp
.if n \{\
.RS 4
.\}
.nf
        \fI[samba_zfs_share]\fR
	\m[blue]\fBvfs objects = zfsacl\fR\m[]
	\m[blue]\fBpath = /test/zfs_mount\fR\m[]
	\m[blue]\fBnfs4: mode = simple\fR\m[]
	\m[blue]\fBnfs4: acedup = merge\fR\m[]
.fi
.if n \{\
.RE
.\}
.SH "VERSION"
.PP
This man page is part of version 4\&.12\&.5 of the Samba suite\&.
.SH "AUTHOR"
.PP
The original Samba software and related utilities were created by Andrew Tridgell\&. Samba is now developed by the Samba Team as an Open Source project similar to the way the Linux kernel is developed\&.


debian/NEWS
samba (2:4.3.7+dfsg-1) unstable; urgency=high

    This Samba security addresses both Denial of Service and Man in
    the Middle vulnerabilities.

    Both of these changes implement new smb.conf options and a number
    of stricter behaviours to prevent Man in the Middle attacks on our
    network services, as a client and as a server.

    Between these changes, compatibility with a large number of older
    software versions has been lost in the default configuration.

    See the release notes in WHATNEW.txt for more information.


    Here are some additional hints how to work around the new stricter default behaviors:

    * As an AD DC server, only Windows 2000 and Samba 3.6 and above as
      a domain member are supported out of the box. Other smb file
      servers as domain members are also fine out of the box.

    * As an AD DC server, with default setting of "ldap server require
      strong auth", LDAP clients connecting over ldaps:// or START_TLS
      will be allowed to perform simple LDAP bind only.

      The preferred configuration for LDAP clients is to use SASL
      GSSAPI directly over ldap:// without using ldaps:// or
      START_TLS.

      To use LDAP with START_TLS and SASL GSSAPI (either Kerberos or
      NTLMSSP) sign/seal protection must be used by the client and
      server should be configured with "ldap server require strong
      auth = allow_sasl_over_tls".

      Consult OpenLDAP documentation how to set sign/seal protection
      in ldap.conf.

      For SSSD client configured with "id_provider = ad" or
      "id_provider = ldap" with "auth_provider = krb5", see
      sssd-ldap(5) manual for details on TLS session handling.

    * As a File Server, compatibility with the Linux Kernel cifs
      client depends on which configuration options are selected, please
      use "sec=krb5(i)" or "sec=ntlmssp(i)", not "sec=ntlmv2".

    * As a file or printer client and as a domain member, out of the
      box compatibility with Samba less than 4.0 and other SMB/CIFS
      servers, depends on support for SMB signing or SMB2 on the
      server, which is often disabled or absent. You may need to
      adjust the "client ipc signing" to "no" in these cases.

    * In case of an upgrade from versions before 4.2.0, you might run
      into problems as a domain member. The out of the box compatibility
      with Samba 3.x domain controllers requires NETLOGON features only
      available in Samba 3.2 and above.
    
    However, all of these can be worked around by setting smb.conf
    options in Samba, see WHATSNEW.txt the 4.2.0 release notes at
    https://www.samba.org/samba/history/samba-4.2.0.html and the Samba
    wiki for details, workarounds and suggested security-improving
    changes to these and other software packages.


    Suggested further improvements after patching:

    It is recommended that administrators set these additional options,
    if compatible with their network environment:

        server signing = mandatory
        ntlm auth = no

    Without "server signing = mandatory", Man in the Middle attacks
    are still possible against our file server and
    classic/NT4-like/Samba3 Domain controller. (It is now enforced on
    Samba's AD DC.) Note that this has heavy impact on the file server
    performance, so you need to decide between performance and
    security. These Man in the Middle attacks for smb file servers are
    well known for decades.

    Without "ntlm auth = no", there may still be clients not using
    NTLMv2, and these observed passwords may be brute-forced easily using
    cloud-computing resources or rainbow tables.

 -- Andrew Bartlett <abartlet+debian@catalyst.net.nz>  Tue, 12 Apr 2016 16:18:57 +1200

samba (2:4.0.10+dfsg-3) unstable; urgency=low

    The SWAT package is no longer available.

    Upstream support for SWAT (Samba Web Administration Tool) was removed in
    samba 4.1.0. As a result, swat is no longer shipped in the Debian Samba
    packages. Unfortunately, there is currently no replacement.

    Details why SWAT has been removed upstream can be found on the
    samba-technical mailing list:

    https://lists.samba.org/archive/samba-technical/2013-February/090572.html

 -- Ivo De Decker <ivo.dedecker@ugent.be>  Tue, 22 Oct 2013 07:52:54 +0200

samba (2:3.6.5-2) unstable; urgency=low

    NSS modules have been split out from libpam-winbind to
    libnss-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libnss-winbind package after upgrading
    from former versions of winbind (for instance from squeeze) or
    from former versions of libpam-winbind.

 -- Christian Perrier <bubulle@debian.org>  Mon, 07 May 2012 22:16:32 +0200

samba (2:3.5.11~dfsg-3) unstable; urgency=low

    PAM modules and NSS modules have been split out from the winbind
    package into libpam-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libpam-winbind package after upgrading
    from former versions of winbind (for instance from squeeze)

 -- Steve Langasek <vorlon@debian.org>  Fri, 21 Oct 2011 20:00:13 +0000

samba (2:3.4.0-1) unstable; urgency=low

    Default passdb backend changed in samba 3.4.0 and above
  
    Beginning with samba 3.4.0, the default setting for "passdb
    backend" changed from "smbpasswd" to "tdbsam".
    
    If your smb.conf file does not have an explicit mention of
    "passdb backend" when upgrading from pre-3.4.0 versions of
    samba, it is likely that users will no longer be able to
    authenticate.
    
    As a consequence of all this, if you're upgrading from lenny
    and have no setting of "passdb backend" in smb.conf, you MUST
    add "passdb backend = smbpasswd" in order to keep your samba
    server's behaviour.
    
    As Debian packages of samba explicitly set "passdb backend = tdbsam"
    by default since etch, very few users should need to modify their

source3/modules/vfs_zfsacl.c
/*
 * Convert ZFS/NFSv4 acls to NT acls and vice versa.
 *
 * Copyright (C) Jiri Sasek, 2007
 * based on the foobar.c module which is copyrighted by Volker Lendecke
 *
 * Many thanks to Axel Apitz for help to fix the special ace's handling
 * issues.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "includes.h"
#include "system/filesys.h"
#include "smbd/smbd.h"
#include "nfs4_acls.h"

#if HAVE_FREEBSD_SUNACL_H
#include "sunacl.h"
#endif

#undef DBGC_CLASS
#define DBGC_CLASS DBGC_VFS

#define ZFSACL_MODULE_NAME "zfsacl"

#define ZFSACL_MODIFY_SET (SMB_ACE4_READ_DATA | SMB_ACE4_READ_ACL \                                       
        | SMB_ACE4_WRITE_DATA | SMB_ACE4_APPEND_DATA | SMB_ACE4_READ_NAMED_ATTRS \                        
        | SMB_ACE4_WRITE_NAMED_ATTRS | SMB_ACE4_EXECUTE | SMB_ACE4_DELETE_CHILD \                         
        | SMB_ACE4_READ_ATTRIBUTES | SMB_ACE4_WRITE_ATTRIBUTES | SMB_ACE4_DELETE \                        
        | SMB_ACE4_SYNCHRONIZE)
#define ZFSACL_READ_SET (SMB_ACE4_READ_DATA | SMB_ACE4_READ_ACL \
        | SMB_ACE4_READ_NAMED_ATTRS | SMB_ACE4_EXECUTE | SMB_ACE4_READ_ATTRIBUTES \
        | SMB_ACE4_SYNCHRONIZE)
#define ZFSACL_WRITE_ONLY_SET (SMB_ACE4_READ_ACL | SMB_ACE4_WRITE_DATA \
        | SMB_ACE4_READ_NAMED_ATTRS | SMB_ACE4_WRITE_NAMED_ATTRS | SMB_ACE4_EXECUTE \
        | SMB_ACE4_READ_ATTRIBUTES | SMB_ACE4_WRITE_ATTRIBUTES | SMB_ACE4_DELETE \
        | SMB_ACE4_SYNCHRONIZE)
#define ZFSACL_BASE_SET (SMB_ACE4_READ_ACL | SMB_ACE4_READ_ATTRIBUTES \
        | SMB_ACE4_READ_NAMED_ATTRS)

static struct SMB4ACL_T *zfsacl_defaultacl(TALLOC_CTX *mem_ctx) 
{
       struct SMB4ACL_T *pacl = NULL;

        /* Owner@ ACE */
        SMB_ACE4PROP_T owner_ace = {
                .flags = SMB_ACE4_ID_SPECIAL,
                .who = {
                        .id = SMB_ACE4_WHO_OWNER,
                },
                .aceType = SMB_ACE4_ACCESS_ALLOWED_ACE_TYPE,
                .aceFlags = 0,
                .aceMask = ZFSACL_READ_SET 
        };

        /* group@ ACE */
        SMB_ACE4PROP_T group_ace = {
                .flags = SMB_ACE4_ID_SPECIAL,
                .who = {
                        .id = SMB_ACE4_WHO_GROUP,
                },
                .aceType = SMB_ACE4_ACCESS_ALLOWED_ACE_TYPE,
                .aceFlags = 0,
                .aceMask = ZFSACL_READ_SET 
        };

        /* everyone@ ACE */
        SMB_ACE4PROP_T everyone_ace = {
                .flags = SMB_ACE4_ID_SPECIAL,
                .who = {
                        .id = SMB_ACE4_WHO_EVERYONE,
                },
                .aceType = SMB_ACE4_ACCESS_ALLOWED_ACE_TYPE,
                .aceFlags = 0,
                .aceMask = ZFSACL_READ_SET 
        };
 
       /* We've converted mode bits to SMB_ACE4PROP_T, add them to ACL */ 
        pacl = smb_create_smb4acl(mem_ctx);
        if (pacl == NULL) {
                DEBUG(0, ("talloc failed\n"));
                errno = ENOMEM;
                return NULL;
        } 
 
        if(smb_add_ace4(pacl, &owner_ace) == NULL) {
                DEBUG(0, ("talloc failed\n"));
                TALLOC_FREE(pacl);
                errno = ENOMEM;
                return NULL;
        }

        if(smb_add_ace4(pacl, &group_ace) == NULL) {
                DEBUG(0, ("talloc failed\n"));
                TALLOC_FREE(pacl);
                errno = ENOMEM;
                return NULL;
        }

        if(smb_add_ace4(pacl, &everyone_ace) == NULL) {
                DEBUG(0, ("talloc failed\n"));
                TALLOC_FREE(pacl);
                errno = ENOMEM;
                return NULL;
        }

        /* 
         * Set DACL_Protected control bit to ensure that Windows Explorer won't try to
         * change permissions on the snapdir when changing them at the root of the share.
         */
        smbacl4_set_controlflags(pacl, SEC_DESC_DACL_PROTECTED|SEC_DESC_SELF_RELATIVE);

        return pacl;
}

/* zfs_get_nt_acl()
 * read the local file's acls and return it in NT form
 * using the NFSv4 format conversion
 */
static NTSTATUS zfs_get_nt_acl_common(struct connection_struct *conn,
				      TALLOC_CTX *mem_ctx,
				      const struct smb_filename *smb_fname,
				      struct SMB4ACL_T **ppacl)
{
	int naces, i;
	ace_t *acebuf;
	struct SMB4ACL_T *pacl;
	bool inherited_present = false;
	SMB_STRUCT_STAT sbuf;
	const SMB_STRUCT_STAT *psbuf = NULL;
	int ret;
	bool is_dir;

	if (VALID_STAT(smb_fname->st)) {
		psbuf = &smb_fname->st;
	}

	if (psbuf == NULL) {
		ret = vfs_stat_smb_basename(conn, smb_fname, &sbuf);
		if (ret != 0) {
			DBG_INFO("stat [%s]failed: %s\n",
				 smb_fname_str_dbg(smb_fname), strerror(errno));
			return map_nt_error_from_unix(errno);
		}
		psbuf = &sbuf;
	}
	is_dir = S_ISDIR(psbuf->st_ex_mode);

	/* read the number of file aces */
	if((naces = acl(smb_fname->base_name, ACE_GETACLCNT, 0, NULL)) == -1) {
		if(errno == ENOSYS) {
			DEBUG(9, ("acl(ACE_GETACLCNT, %s): Operation is not "
				  "supported on the filesystem where the file "
				  "reside\n", smb_fname->base_name));
                        if(lp_parm_bool(conn->params->service, "zfsacl", "expose_snapdir", false)) {
                                *ppacl = zfsacl_defaultacl(mem_ctx);
                                return NT_STATUS_OK;
			}
		} else {
			DEBUG(9, ("acl(ACE_GETACLCNT, %s): %s ", smb_fname->base_name,
					strerror(errno)));
		}
		return map_nt_error_from_unix(errno);
	}
	/* allocate the field of ZFS aces */
	mem_ctx = talloc_tos();
	acebuf = (ace_t *) talloc_size(mem_ctx, sizeof(ace_t)*naces);
	if(acebuf == NULL) {
		return NT_STATUS_NO_MEMORY;
	}
	/* read the aces into the field */
	if(acl(smb_fname->base_name, ACE_GETACL, naces, acebuf) < 0) {
		DEBUG(9, ("acl(ACE_GETACL, %s): %s ", smb_fname->base_name,
				strerror(errno)));
		return map_nt_error_from_unix(errno);
	}
	/* create SMB4ACL data */
	if((pacl = smb_create_smb4acl(mem_ctx)) == NULL) {
		return NT_STATUS_NO_MEMORY;
	}
	for(i=0; i<naces; i++) {
		SMB_ACE4PROP_T aceprop;

		aceprop.aceType  = (uint32_t) acebuf[i].a_type;
		aceprop.aceFlags = (uint32_t) acebuf[i].a_flags;
		aceprop.aceMask  = (uint32_t) acebuf[i].a_access_mask;
		aceprop.who.id   = (uint32_t) acebuf[i].a_who;

 		/*
 		 * Test whether ACL contains any ACEs with the
 		 * inherited flag set. We use this to determine whether
 		 * to set DACL_PROTECTED in the security descriptor.
 		 */
 		if(aceprop.aceFlags & ACE_INHERITED_ACE) {
 			inherited_present = true;
		}
		
		/*
		 * Windows clients expect SYNC on acls to correctly allow
		 * rename, cf bug #7909. But not on DENY ace entries, cf bug
		 * #8442.
		 */
		if (aceprop.aceType == SMB_ACE4_ACCESS_ALLOWED_ACE_TYPE) {
			aceprop.aceMask |= SMB_ACE4_SYNCHRONIZE;
		}

		if (is_dir && (aceprop.aceMask & SMB_ACE4_ADD_FILE)) {
			aceprop.aceMask |= SMB_ACE4_DELETE_CHILD;
		}

		if(aceprop.aceFlags & ACE_OWNER) {
			aceprop.flags = SMB_ACE4_ID_SPECIAL;
			aceprop.who.special_id = SMB_ACE4_WHO_OWNER;
		} else if(aceprop.aceFlags & ACE_GROUP) {
			aceprop.flags = SMB_ACE4_ID_SPECIAL;
			aceprop.who.special_id = SMB_ACE4_WHO_GROUP;
		} else if(aceprop.aceFlags & ACE_EVERYONE) {
			aceprop.flags = SMB_ACE4_ID_SPECIAL;
			aceprop.who.special_id = SMB_ACE4_WHO_EVERYONE;
		} else {
			aceprop.flags	= 0;
		}
		if(smb_add_ace4(pacl, &aceprop) == NULL)
			return NT_STATUS_NO_MEMORY;
	}
	
	/*
 	 * If the ACL doesn't contain any inherited ACEs, then set DACL_PROTECTED 
 	 * in the security descriptor using smb4acl4_set_control_flags() from
 	 * source3/modules/nfs4_acls.c. This makes it so that the "Disable 
 	 * Inheritance" button works in Windows Explorer and prevents resulting 
 	 * ACL from auto-inheriting ACL changes in parent directory.
 	 */
 	if (!inherited_present 
	    && lp_parm_bool(conn->params->service, "zfsacl", "map_dacl_protected", false)){
 		smbacl4_set_controlflags(pacl, SEC_DESC_DACL_PROTECTED|SEC_DESC_SELF_RELATIVE);
	}

	*ppacl = pacl;
	return NT_STATUS_OK;
}

/* call-back function processing the NT acl -> ZFS acl using NFSv4 conv. */
static bool zfs_process_smbacl(vfs_handle_struct *handle, files_struct *fsp,
			       struct SMB4ACL_T *smbacl)
{
	int naces = smb_get_naces(smbacl), i;
	ace_t *acebuf;
	struct SMB4ACE_T *smbace;
	TALLOC_CTX	*mem_ctx;
	bool have_special_id = false;

	/* allocate the field of ZFS aces */
	mem_ctx = talloc_tos();
	acebuf = (ace_t *) talloc_size(mem_ctx, sizeof(ace_t)*naces);
	if(acebuf == NULL) {
		errno = ENOMEM;
		return False;
	}
	/* handle all aces */
	for(smbace = smb_first_ace4(smbacl), i = 0;
			smbace!=NULL;
			smbace = smb_next_ace4(smbace), i++) {
		SMB_ACE4PROP_T *aceprop = smb_get_ace4(smbace);

		acebuf[i].a_type        = aceprop->aceType;
		acebuf[i].a_flags       = aceprop->aceFlags;
		acebuf[i].a_access_mask = aceprop->aceMask;
		/* SYNC on acls is a no-op on ZFS.
		   See bug #7909. */
		acebuf[i].a_access_mask &= ~SMB_ACE4_SYNCHRONIZE;
		acebuf[i].a_who         = aceprop->who.id;
		if(aceprop->flags & SMB_ACE4_ID_SPECIAL) {
			switch(aceprop->who.special_id) {
			case SMB_ACE4_WHO_EVERYONE:
				acebuf[i].a_flags |= ACE_EVERYONE;
				break;
			case SMB_ACE4_WHO_OWNER:
				acebuf[i].a_flags |= ACE_OWNER;
				break;
			case SMB_ACE4_WHO_GROUP:
				acebuf[i].a_flags |= ACE_GROUP|ACE_IDENTIFIER_GROUP;
				break;
			default:
				DEBUG(8, ("unsupported special_id %d\n", \
					aceprop->who.special_id));
				continue; /* don't add it !!! */
			}
			have_special_id = true;
		}
	}

	if (!have_special_id
	    && lp_parm_bool(fsp->conn->params->service, "zfsacl",
			    "denymissingspecial", false)) {
		errno = EACCES;
		return false;
	}

	SMB_ASSERT(i == naces);

	/* store acl */
	if(acl(fsp->fsp_name->base_name, ACE_SETACL, naces, acebuf)) {
		if(errno == ENOSYS) {
			DEBUG(9, ("acl(ACE_SETACL, %s): Operation is not "
				  "supported on the filesystem where the file "
				  "reside", fsp_str_dbg(fsp)));
		} else {
			DEBUG(9, ("acl(ACE_SETACL, %s): %s ", fsp_str_dbg(fsp),
				  strerror(errno)));
		}
		return 0;
	}

	return True;
}

/* zfs_set_nt_acl()
 * set the local file's acls obtaining it in NT form
 * using the NFSv4 format conversion
 */
static NTSTATUS zfs_set_nt_acl(vfs_handle_struct *handle, files_struct *fsp,
			   uint32_t security_info_sent,
			   const struct security_descriptor *psd)
{
        return smb_set_nt_acl_nfs4(handle, fsp, NULL, security_info_sent, psd,
				   zfs_process_smbacl);
}

static NTSTATUS zfsacl_fget_nt_acl(struct vfs_handle_struct *handle,
				   struct files_struct *fsp,
				   uint32_t security_info,
				   TALLOC_CTX *mem_ctx,
				   struct security_descriptor **ppdesc)
{
	struct SMB4ACL_T *pacl;
	NTSTATUS status;
	TALLOC_CTX *frame = talloc_stackframe();

	status = zfs_get_nt_acl_common(handle->conn, frame,
				       fsp->fsp_name, &pacl);
	if (!NT_STATUS_IS_OK(status)) {
		TALLOC_FREE(frame);
		return status;
	}

	status = smb_fget_nt_acl_nfs4(fsp, NULL, security_info, mem_ctx,
				      ppdesc, pacl);
	TALLOC_FREE(frame);
	return status;
}

static NTSTATUS zfsacl_get_nt_acl(struct vfs_handle_struct *handle,
				const struct smb_filename *smb_fname,
				uint32_t security_info,
				TALLOC_CTX *mem_ctx,
				struct security_descriptor **ppdesc)
{
	struct SMB4ACL_T *pacl;
	NTSTATUS status;
	TALLOC_CTX *frame = talloc_stackframe();

	status = zfs_get_nt_acl_common(handle->conn, frame, smb_fname, &pacl);
	if (!NT_STATUS_IS_OK(status)) {
		TALLOC_FREE(frame);
		return status;
	}

	status = smb_get_nt_acl_nfs4(handle->conn,
					smb_fname,
					NULL,
					security_info,
					mem_ctx,
					ppdesc,
					pacl);
	TALLOC_FREE(frame);
	return status;
}

static NTSTATUS zfsacl_fset_nt_acl(vfs_handle_struct *handle,
			 files_struct *fsp,
			 uint32_t security_info_sent,
			 const struct security_descriptor *psd)
{
	return zfs_set_nt_acl(handle, fsp, security_info_sent, psd);
}

/* nils.goroll@hamburg.de 2008-06-16 :

   See also
   - https://bugzilla.samba.org/show_bug.cgi?id=5446
   - http://bugs.opensolaris.org/view_bug.do?bug_id=6688240

   Solaris supports NFSv4 and ZFS ACLs through a common system call, acl(2)
   with ACE_SETACL / ACE_GETACL / ACE_GETACLCNT, which is being wrapped for
   use by samba in this module.

   As the acl(2) interface is identical for ZFS and for NFS, this module,
   vfs_zfsacl, can not only be used for ZFS, but also for sharing NFSv4
   mounts on Solaris.

   But while "traditional" POSIX DRAFT ACLs (using acl(2) with SETACL
   / GETACL / GETACLCNT) fail for ZFS, the Solaris NFS client
   implemets a compatibility wrapper, which will make calls to
   traditional ACL calls though vfs_solarisacl succeed. As the
   compatibility wrapper's implementation is (by design) incomplete,
   we want to make sure that it is never being called.

   As long as Samba does not support an exiplicit method for a module
   to define conflicting vfs methods, we should override all conflicting
   methods here.

   For this to work, we need to make sure that this module is initialised
   *after* vfs_solarisacl

   Function declarations taken from vfs_solarisacl
*/

static SMB_ACL_T zfsacl_fail__sys_acl_get_file(vfs_handle_struct *handle,
					const struct smb_filename *smb_fname,
					SMB_ACL_TYPE_T type,
					TALLOC_CTX *mem_ctx)
{
	return (SMB_ACL_T)NULL;
}

static SMB_ACL_T zfsacl_fail__sys_acl_get_fd(vfs_handle_struct *handle,
					     files_struct *fsp,
					     TALLOC_CTX *mem_ctx)
{
	return (SMB_ACL_T)NULL;
}

static int zfsacl_fail__sys_acl_set_file(vfs_handle_struct *handle,
					 const struct smb_filename *smb_fname,
					 SMB_ACL_TYPE_T type,
					 SMB_ACL_T theacl)
{
	return -1;
}

static int zfsacl_fail__sys_acl_set_fd(vfs_handle_struct *handle,
				       files_struct *fsp,
				       SMB_ACL_T theacl)
{
	return -1;
}

static int zfsacl_fail__sys_acl_delete_def_file(vfs_handle_struct *handle,
			const struct smb_filename *smb_fname)
{
	return -1;
}

static int zfsacl_fail__sys_acl_blob_get_file(vfs_handle_struct *handle,
			const struct smb_filename *smb_fname,
			TALLOC_CTX *mem_ctx,
			char **blob_description,
			DATA_BLOB *blob)
{
	return -1;
}

static int zfsacl_fail__sys_acl_blob_get_fd(vfs_handle_struct *handle, files_struct *fsp, TALLOC_CTX *mem_ctx, char **blob_description, DATA_BLOB *blob)
{
	return -1;
}

/* VFS operations structure */

static struct vfs_fn_pointers zfsacl_fns = {
	.sys_acl_get_file_fn = zfsacl_fail__sys_acl_get_file,
	.sys_acl_get_fd_fn = zfsacl_fail__sys_acl_get_fd,
	.sys_acl_blob_get_file_fn = zfsacl_fail__sys_acl_blob_get_file,
	.sys_acl_blob_get_fd_fn = zfsacl_fail__sys_acl_blob_get_fd,
	.sys_acl_set_file_fn = zfsacl_fail__sys_acl_set_file,
	.sys_acl_set_fd_fn = zfsacl_fail__sys_acl_set_fd,
	.sys_acl_delete_def_file_fn = zfsacl_fail__sys_acl_delete_def_file,
	.fget_nt_acl_fn = zfsacl_fget_nt_acl,
	.get_nt_acl_fn = zfsacl_get_nt_acl,
	.fset_nt_acl_fn = zfsacl_fset_nt_acl,
};

NTSTATUS vfs_zfsacl_init(TALLOC_CTX *);
NTSTATUS vfs_zfsacl_init(TALLOC_CTX *ctx)
{
	return smb_register_vfs(SMB_VFS_INTERFACE_VERSION, "zfsacl",
				&zfsacl_fns);
}
